airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/HeartbeatTimeoutChaperone.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers.internal;
+
+import static java.lang.Thread.sleep;
+
+import io.airbyte.featureflag.FeatureFlagClient;
+import io.airbyte.featureflag.ShouldFailSyncIfHeartbeatFailure;
+import io.airbyte.featureflag.Workspace;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link HeartbeatTimeoutChaperone} takes in an arbitrary runnable and a heartbeat monitor. It
+ * runs each in separate threads. If the heartbeat monitor thread completes before the runnable,
+ * that means that the heartbeat has stopped. If this occurs the chaperone cancels the runnable
+ * thread and then throws an exception. If the runnable thread completes first, the chaperone
+ * cancels the heartbeat and then returns.
+ *
+ * This allows us to run an arbitrary runnable that we can kill if a heartbeat stops. This is useful
+ * in cases like the platform reading from the source. The thread that reads from the source is
+ * allowed to run as long as the heartbeat from the sources is fresh.
+ */
+public class HeartbeatTimeoutChaperone {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HeartbeatTimeoutChaperone.class);
+
+  public static final Duration DEFAULT_TIMEOUT_CHECK_DURATION = Duration.ofMinutes(1);
+
+  private final HeartbeatMonitor heartbeatMonitor;
+  private final Duration timeoutCheckDuration;
+  private final FeatureFlagClient featureFlagClient;
+  private final UUID workspaceId;
+
+  private boolean hasFailed = false;
+
+  public HeartbeatTimeoutChaperone(final HeartbeatMonitor heartbeatMonitor,
+                                   final Duration timeoutCheckDuration,
+                                   final FeatureFlagClient featureFlagClient,
+                                   final UUID workspaceId) {
+    this.timeoutCheckDuration = timeoutCheckDuration;
+
+    LOGGER.info("Starting source heartbeat check. Will check every {} minutes.", timeoutCheckDuration.toMinutes());
+
+    this.heartbeatMonitor = heartbeatMonitor;
+    this.featureFlagClient = featureFlagClient;
+    this.workspaceId = workspaceId;
+  }
+
+  public Runnable runWithHeartbeatThread(final Runnable runnable, final ExecutorService executorService) {
+    return () -> {
+      final CompletableFuture<Void> runnableFuture = CompletableFuture.runAsync(runnable, executorService);
+      final CompletableFuture<Void> heartbeatFuture = CompletableFuture.runAsync(this::monitor, executorService);
+
+      try {
+        CompletableFuture.anyOf(runnableFuture, heartbeatFuture).get();
+      } catch (final InterruptedException e) {
+        LOGGER.error("Heartbeat chaperone thread was interrupted.", e);
```

Comment:
If runnableFuture end up throwing an interrupted exception, I think we'd end up logging the wrong information here.
Should we reuse the same logic as l80 instead?

Reply:
I'm donb't think so. An interruption exception doesn't mean that there was a timeout. I wonder if that might happen when we cancel a sync which is not a timeout error.

Reply:
Got it